crates/ruff_linter/src/rules/ruff/rules/operation_on_closed_io.rs

```diff
@@ -0,0 +1,179 @@
+use crate::checkers::ast::Checker;
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, ViolationMetadata};
+use ruff_python_ast::AnyNodeRef;
+use ruff_python_semantic::analyze::typing;
+use ruff_python_semantic::{Binding, BindingKind, NodeId, SemanticModel};
+use ruff_text_size::{Ranged, TextRange};
+
+/// ## What it does
+/// Checks for usages of IO operation methods of context variables
+/// outside of the original `with` statement.
+///
+/// ## Why is this bad?
+/// Such operations will raise `ValueError: I/O operation on closed file` at runtime.
+///
+/// ## Example
+///
+/// ```python
+/// with open(".txt") as f:
+///     f.read()
+///
+/// with open(".md", "w"):
+///     f.write("")
+/// ```
+///
+/// Use instead:
+///
+/// ```python
+/// with open(".txt") as f:
+///     f.read()
+///
+/// with open(".md", "w") as f:
+///     f.write("")
+/// ```
+#[derive(ViolationMetadata)]
+pub(crate) struct OperationOnClosedIO;
+
+impl Violation for OperationOnClosedIO {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        "IO operation performed on closed IO object".to_string()
+    }
+}
+
+/// RUF050
+pub(crate) fn operation_on_closed_io(
+    checker: &Checker,
+    binding: &Binding,
+) -> Option<Vec<Diagnostic>> {
+    if !matches!(&binding.kind, BindingKind::WithItemVar) {
+        return None;
+    }
+
+    let semantic = checker.semantic();
+    let with = binding.statement(semantic)?.as_with_stmt()?;
+
+    if !typing::is_io_base(binding, semantic) {
+        return None;
+    }
+
+    let mut diagnostics = vec![];
+
+    for reference_id in binding.references() {
+        let reference = semantic.reference(reference_id);
+
+        if reference.end() <= with.end() {
+            continue;
+        }
+
+        let Some(expression_id) = reference.expression_id() else {
+            continue;
+        };
+
+        let Some(range) = method_reference_range(expression_id, semantic)
+            .or_else(|| contains_check_range(expression_id, semantic))
+            .or_else(|| for_loop_target_in_iter_range(expression_id, semantic))
+        else {
+            continue;
+        };
+
+        let diagnostic = Diagnostic::new(OperationOnClosedIO, range);
+
+        diagnostics.push(diagnostic);
+    }
+
+    Some(diagnostics)
+}
+
+// `f.write(...)`
+fn method_reference_range(expression_id: NodeId, semantic: &SemanticModel) -> Option<TextRange> {
```

Comment:
I think I now have a better understanding of the method names. How about `io_func_range`, `in_file_range`, and `io_in_for_loop_range`?